accepted/0000-workspaces.md

```diff
@@ -0,0 +1,480 @@
+# npm workspaces
+
+## Summary
+
+Add a set of features to the npm cli that provide support to managing multiple packages from within a singular top-level, root package.
+
+## Motivation
+
+This feature has been requested by the community for a long time. The primary motivation behind this RFC is to fully realize a set of features/functionality to support managing multiple packages that may or may not be used together.
+
+The name “workspaces” is already established in the community with both Yarn and Pnpm implementing similar features under that same name so we chose to reuse it for the sake of simplicity to the larger community involved.
+
+## Detailed Explanation
+
+After sourcing feedback from the community, there are 2 major implementations/changes required in the cli in order to provide the feature set that would enable a better management of nested packages.
+- Install: In a workspaces setup users expect to be able to install all nested packages and perform the associated lifecycle scripts from the root-level, it should also be aware of sibling packages that depend on one another and link them appropriately
+- Run some npm subcommands in the context of each nested workspace
+
+## Rationale and Alternatives
+
+First and foremost there’s the alternative of leaving the problem set for userland to solve, there’s already the very popular project Lerna that provides some of these features.
+
+Also available is the alternative of supporting only the install (or bootstrap as Lerna names it) aspect of this proposal, following a less feature-rich approach but one that would still enable the basic goal of improving the user experience of managing multiple child packages but from all the feedback collected during the research phase of this RFC, this alternative is much less desirable to the community of maintainers involved.
+
+## Implementation
+
+### 1. Make npm workspace-aware
+
+We're following the lead of Yarn in supporting the `workspaces` `package.json` field which defines a list of paths, each of these paths may be a workspace itself but it also support globs.
+
+`package.json` example:
+
+```
+{
+    "name": "workspace-example",
+    "version": "1.0.0",
+    "workspaces": {
```

Comment:
should we also support the singular form `workspace` ?

Reply:
oof, let's please not repeat "bundleDependencies" vs "bundledDependencies" :-)

Reply:
Isn't a singular `workspace` representing a collection of packages? Calling each nested package a workspace seems to be misleading.

Reply:
thanks @raymondfeng I did a round of cleaning up the RFC in which I defined a **Dictionary** section and reused the exact same words everywhere (to which people can refer at the dictionary section of the RFC) in order to make the entire conversation more straightforward

to answer to your question, the collection of packages are named **workspaces** (so each package of this collection is referred to as a **workspace**), the root-level package in which you define the configurations for working with **npm workspaces** is the **Top-level workspace**

from all the documentation I read across Yarn and pnpm, this terminology seems to reflect better the relationship the community is already using (feel free to correct me if I'm wrong)

Reply:
@ruyadorno pnpm is using "workspace" as a set of projects. I don't understand why Yarn decided to use "workspace" as a synonym for package (or project).

I also did a small poll. Most of pnpm users believe that a workspace should be a set of projects: https://twitter.com/ZoltanKochan/status/1214223464363675649

Also, pnpm stores the package globs in a separate file called `pnpm-workspace.yaml`, which defines the root of the workspace. It is much easier to find the root of the workspace by looking for `pnpm-workspace.yaml` vs looking for a package.json with a workspace field. A pnpm workspace may even be without a package.json in the root.

Reply:
Seems like even @isaacs agreed with the definition used by pnpm: https://twitter.com/izs/status/1214305689977253888